tee-worker/litentry/core/vc-task/receiver/src/lib.rs

```diff
@@ -154,30 +162,122 @@ where
 	N: AccessNodeMetadata + Send + Sync + 'static,
 	N::MetadataType: NodeMetadataTrait,
 {
-	let start_time = Instant::now();
-	let enclave_shielding_key = context
-		.shielding_key
-		.retrieve_key()
-		.map_err(|e| format!("Failed to retrieve shielding key: {:?}", e))?;
-	let tcs = request
+	let request = &mut vc_req.request;
+	let sender = &vc_req.sender;
+	let enclave_shielding_key = match context.shielding_key.retrieve_key() {
+		Ok(value) => value,
+		Err(e) => {
+			send_vc_response(
+				sender,
+				context,
+				Err(format!("Failed to retrieve shielding key: {:?}", e)),
+			);
+			return
+		},
+	};
+	let tcs = match request
 		.decrypt(Box::new(enclave_shielding_key))
 		.ok()
 		.and_then(|v| TrustedOperation::<TrustedCallSigned, Getter>::decode(&mut v.as_slice()).ok())
 		.and_then(|top| top.to_call().cloned())
-		.ok_or_else(|| "Failed to decode payload".to_string())?;
-
+	{
+		Some(tcs) => tcs,
+		None => {
+			send_vc_response(sender, context, Err("Failed to decode payload".to_string()));
+			return
+		},
+	};
 	let mrenclave = match context.ocall_api.get_mrenclave_of_self() {
 		Ok(m) => m.m,
-		Err(_) => return Err("Failed to get mrenclave".to_string()),
+		Err(_) => {
+			send_vc_response(sender, context, Err("Failed to get mrenclave".to_string()));
+			return
+		},
 	};
+	if !tcs.verify_signature(&mrenclave, &request.shard) {
+		send_vc_response(sender, context, Err("Failed to verify sig".to_string()));
+		return
+	}
+	if let TrustedCall::request_vc(..) = tcs.call {
+		let response = process_single_request(
+			request.shard,
+			context.clone(),
+			extrinsic_factory,
+			node_metadata_repo,
+			tc_sender,
+			tcs.call.clone(),
+			1,
+		);
+		send_vc_response(sender, context, response);
+	} else if let TrustedCall::batch_request_vc(signer, who, assertions, maybe_key, req_ext_hash) =
+		tcs.call
+	{
+		let pool = ThreadPool::new(4);
+		let len = assertions.len() as u8;
+		for assertion in &assertions {
+			let new_call = TrustedCall::request_vc(
+				signer.clone(),
+				who.clone(),
+				assertion.clone(),
+				maybe_key,
+				req_ext_hash,
+			);
+			let context_pool = context.clone();
+			let extrinsic_factory_pool = extrinsic_factory.clone();
+			let node_metadata_repo_pool = node_metadata_repo.clone();
+			let tc_sender_pool = tc_sender.clone();
+			let shard = request.shard;
+			let sender_clone = sender.clone();
+
+			pool.execute(move || {
+				let response = process_single_request(
```

Comment:
So every single request we are submitting a callback extrinsic upon success, Can we batch them as well for the batch request-vc? wdyt? Could be useful when we implement callback extrinsics for failure as well.

Reply:
Thought about it before. But there the use case would easily result in a "select all" option from FE; then a big vector of assertions (e.g. 30) come inside. If we combine all result together, that would too big as one result. That's the reason to split it and send back result separately.
Otherwise, it would much simpler only to extend the old arch to process vector of assertions, gather result together, and return. Only need to touch one file is enough.

Reply:
The FE is not using callback extrinsics for anything, It's only meant for logging. We shouldn't allow a big vector of assertion, which is my other comment.